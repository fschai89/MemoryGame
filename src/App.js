'use strict';
import React, {useState, useEffect, useCallback} from 'react';
import {
  StyleSheet,
  TextInput,
  View,
  Text,
  SafeAreaView,
  Pressable,
  Alert,
  Dimensions,
  TouchableOpacity,
} from 'react-native';
import FlipCard from './reuseable/FlipCard';
import Button from './reuseable/Button';
import {moderateUIScale} from './utility/UIScale';

import {useDimensions, shuffleArray, generateNumber} from './utility/Utils';
import styles from './styles/appStyle';

export default function App() {
  const [dimensions, onLayout] = useDimensions();
  const {width, height} = Dimensions.get('window');

  const NUMBER_OF_CARD = 12;

  const [CARD_PAIRS_VALUE, setCards] = useState(generateNumber(NUMBER_OF_CARD));
  const [steps, setSteps] = useState(0);
  const [firstCardOpened, setFirstCardOpened] = useState(undefined);
  const [secondCardOpened, setSecondCardOpened] = useState(undefined);
  const [isFlip, setIsFlip] = useState(false);

  const restart = () => {
    CARD_PAIRS_VALUE.map((card) => {
      card.show = false;
    });
    setCards([...CARD_PAIRS_VALUE]);
    setSteps(0);
    setFirstCardOpened(undefined);
    setSecondCardOpened(undefined);

    setTimeout(() => {
      let newCards = generateNumber(NUMBER_OF_CARD);
      setCards(newCards);
    }, 500);
  };

  useEffect(() => {
    if (firstCardOpened && secondCardOpened) {
      if (firstCardOpened.value === secondCardOpened.value) {
        setFirstCardOpened(undefined);
        setSecondCardOpened(undefined);
      } else {
        setTimeout(() => {
          CARD_PAIRS_VALUE.map((card) => {
            if (
              card.id == firstCardOpened.id ||
              card.id == secondCardOpened.id
            ) {
              card.show = false;
            }
          });
          setCards([...CARD_PAIRS_VALUE]);
          setFirstCardOpened(undefined);
          setSecondCardOpened(undefined);
        }, 1000);
      }
    }
  }, [firstCardOpened, secondCardOpened]);

  useEffect(() => {
    let allShow = CARD_PAIRS_VALUE.every((card) => card.show === true);
    if (allShow) {
      Alert.alert('Congratulations!', `You win this game by ${steps} steps!`, [
        {
          text: 'Try another round',
          onPress: () => restart(),
        },
      ]);
    }
  }, [CARD_PAIRS_VALUE]);

  const openCard = (card) => {
    if (!firstCardOpened) {
      setSteps(steps + 1);
      card.show = true;
      setFirstCardOpened(card);
    }
    if (firstCardOpened && card != firstCardOpened && !secondCardOpened) {
      setSteps(steps + 1);
      card.show = true;
      setSecondCardOpened(card);
    }

    setCards([...CARD_PAIRS_VALUE]);
  };

  const cardHeight = dimensions
    ? dimensions.height / 4 - moderateUIScale(10)
    : 0;
  const cardWidth = dimensions ? dimensions.width / 3 - moderateUIScale(15) : 0;

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.headerContainer}>
        <View style={styles.headerLeft}>
          <Button title="Restart" onPress={() => restart()} />
        </View>
        <View style={styles.headerRight}>
          <Text style={styles.font1}>Steps:</Text>
          <Text style={styles.font2}>{steps}</Text>
        </View>
      </View>
      <View style={styles.cardsContainer} onLayout={onLayout}>
        {dimensions &&
          CARD_PAIRS_VALUE.map((card) => {
            return (
              <Pressable
                style={{
                  height: cardHeight,
                  width: cardWidth,
                  marginBottom: moderateUIScale(10),
                }}
                key={card.id}
                onPress={() => openCard(card)}>
                <FlipCard
                  isFlipped={card.show}
                  frontView={
                    <View
                      style={[
                        styles.frontCard,
                        {
                          height: cardHeight,
                          width: cardWidth,
                        },
                      ]}>
                      <Text
                        style={{
                          color: '#fff',
                          fontSize: moderateUIScale(30),
                          fontWeight: 'bold',
                        }}>
                        ?
                      </Text>
                    </View>
                  }
                  backView={
                    <View
                      style={[
                        styles.backCard,
                        {
                          height: cardHeight,
                          width: cardWidth,
                        },
                      ]}>
                      <Text style={styles.font3}>{card.value}</Text>
                    </View>
                  }
                />
              </Pressable>
            );
          })}
      </View>
    </SafeAreaView>
  );
}
