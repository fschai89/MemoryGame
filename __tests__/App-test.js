/**
 * @format
 */

import 'react-native';
import React from 'react';
import App from '../src/App';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  renderer.create(<App />);
});

test('App snapshot', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('<App />', () => {
  it('has 2 child', () => {
    const tree = renderer.create(<App />).toJSON();
    expect(tree.children.length).toBe(2);
  });
});
