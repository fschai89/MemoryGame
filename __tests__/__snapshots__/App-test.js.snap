// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "rgba(0,0,0,0.7)",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "maxHeight": 84,
        "paddingHorizontal": 21,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "#3576E1",
            "borderColor": "#3576E1",
            "borderRadius": 7,
            "borderWidth": 1,
            "opacity": 1,
            "paddingHorizontal": 14,
            "paddingVertical": 7,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#fff",
                "fontSize": 28,
              },
              Object {
                "color": "#fff",
                "textAlign": "center",
              },
            ]
          }
        >
          Restart
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "flex-end",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 35,
          }
        }
      >
        Steps:
      </Text>
      <Text
        style={
          Object {
            "color": "#3576E1",
            "fontSize": 42,
            "paddingLeft": 7,
          }
        }
      >
        0
      </Text>
    </View>
  </View>
  <View
    onLayout={[Function]}
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "flexWrap": "wrap",
        "justifyContent": "space-between",
        "paddingHorizontal": 21,
      }
    }
  />
</RCTSafeAreaView>
`;
